Remove APIs passing model data through memory

yunfengzhou-hub opened the pull request:
This PR removes Clink APIs that transmit model data as string or byte array through memory. Reason is as explained in #10.

Usages of these APIs in Java codes and in C++ tests are changed to load model data from files.

yunfengzhou-hub committed:
Remove APIs passing model data through memory

lindong28 commented on java-lib/src/main/java/org/flinkextended/clink/feature/onehotencoder/ClinkOneHotEncoderModel.java:
```diff
@@ -181,13 +187,24 @@ public void close() throws Exception {
+        File tmpDir = Files.createTempDirectory("ClinkOneHotEncoderModel").toFile();
+        String tmpDirStr = tmpDir.getAbsolutePath();
+
+        ParamUtils.saveMetadata(paramMap, ClinkOneHotEncoderModel.class, tmpDirStr);
```
`ClinkOneHotEncoderModel::save(...)` also saves the metadata and model data. Could `loadCppModel(...)` re-use the `save(...)` instead of duplicating the code here to save metadata and model data?

yunfengzhou-hub replied:
I agree that we should try to reduce duplication in our programs. `ClinkOneHotEncoderModel::save(...)` only constructs the stream graph to save model data, while `loadCppModel()` actually completes the job, so we cannot directly use `ClinkOneHotEncoderModel::save(...)` in `loadCppModel()`. Instead, I tried changing the implementation of `loadCppModel` to use `ByteArrayEncoder()` and `OneHotEncoderProtobufUtils.getModelDataByteArray()`, which have been used in `ClinkOneHotEncoderModel::save(...)`, and this might help to achieve the goal to reduce duplication.

During the fix of this problem I also found and fixed another bug in the saving process: while `loadCppModel()` and `OneHotEncoderModel::load` only saves or reads the model data, `ClinkOneHotEncoderModel::save(...)` saves both the length of model data and the model data itself. I unified them to `ClinkOneHotEncoderModel::save(...)`'s practice. You may notice the related change in `one_hot_encoder.cc`.

lindong28 replied:
Could we change `ClinkOneHotEncoderModel::save(...)` to use `ParamUtils.saveMetadata(...)` as well for better consistency?

Would it be better to rename `ParamUtils` to `ClinkReadWriteUtils` so that the relationship between these two utils is more self-explanatory?

Suppose user runs the following code, the current PR will save and read the entire model data again during `map(..)`, which is inefficient when the model data size is large. Could we avoid this overhead, e.g. load the cppModel during `ClinkOneHotEncoderModel::load(...)`?

```
modelA.save(path_1);
ClinkOneHotEncoderModel modelB = ClinkOneHotEncoderModel::load(path_1)
Table outputTable = modeB.map(...)
```

yunfengzhou-hub replied:
I agree to the points to make `ClinkOneHotEncoderModel::save(...)` use `ParamUtils.saveMetadata(...)` and to rename `ParamUtils` to `ClinkReadWriteUtils`.

As for the third point, `ClinkOneHotEncoderModel` should be able to judge whether itself is loaded from certain file path, and if it is true, avoid re-saving the model data. I'll make the change in this direction.

lindong28 replied:
Has this issue been resolved? It seems that the PR will still load model twice if user runs the code snippet described above?

yunfengzhou-hub replied:
I thought it has been resolved as the program has been able to avoid re-saving model data if it has been saved before. The two loading process might be inevitable: one is invoked in Java program as to support `Model.getModelData()` method, the other is invoked in C++ program to initialize Clink operator. Given that the target of this PR is to remove APIs that transmit model data through memory, we cannot load the data just once and share it among Java and C++ programs.

yunfengzhou-hub replied:
As discussed offline, Given that loading model data into C++ operators is inevitable, I'll change the implementation so that `ClinkOneHotEncoderModel` will not load model data into Java memory space unless APIs that require this information, i.e. `getModelData()`, is invoked.

lindong28 commented on java-lib/src/main/java/org/flinkextended/clink/jna/ClinkJna.java:
```diff
@@ -34,22 +34,16 @@
      *
      * @param vector A reference to the {@link SparseVectorJna} object.
      */
-    // TODO: Automatically free C++ objects to avoid memory leak and improve usability.
```
`ClinkJna.INSTANCE.OneHotEncoderModel_transform` invokes the underlying C++ method `OneHotEncoderModel_transform`. And the C++ object SparseVectorJNA returned the C++ method corresponds to the Java object `SparseVectorJna.ByReference` returned by the Java `OneHotEncoderModel_transform`.

Ideally we would like the C++ object to be automatically destructed when the corresponding Java object is destructed. In theory this should be feasible. We just don't know whether this is actually achievable with the JNA library.

Would it still be useful to keep this TODO and investigate it later?

yunfengzhou-hub replied:
I agree. I'll add back this TODO.

lindong28 commented on lib/feature/one_hot_encoder.cc:
```diff
@@ -91,16 +72,19 @@ OneHotEncoderModel::load(const std::string &path, tfrt::HostContext *host) {
+  if (!model->model_data_.ParseFromIstream(&model_data_input)) {
     return tfrt::MakeStringError(
-        "Failed to load OneHotEncoderModel: invalid model data file " + path +
+        "Failed to load OneHotEncoderModel: Invalid model data file " + path +
```
String concatenation using the `+` operator could be inefficient due to the reasons explained in [1]. Given that we are already using `MakeStringError`, it seems better to use the following style. This style would also be useful in the future when we use non-string value in the error message.

Same for other usages of `MakeStringError`.

```
return tfrt::MakeStringError(
    "Failed to load OneHotEncoderModel: Invalid model data file ", path, "/data/" + model_data_filename);
```

[1] https://stackoverflow.com/questions/611263/efficient-string-concatenation-in-c

yunfengzhou-hub replied:
OK. I'll make the change to all usages of `MakeStringError`.

Apart from `MakeStringError`, there are also many other usages of string concatenation. I also changed all these to `tfrt::StrCat`, which is how `MakeStringError` concatenate the arguments while preserving efficiency.

lindong28 reviewed (commented):
Thanks for the PR. Left some comments below.

yunfengzhou-hub commented:
Thanks for the comments. I have fixed the problems pointed out in the comments and left responses as above.

lindong28 commented on cpp_tests/include/clink/cpp_tests/test_util.h:
```diff
@@ -78,12 +78,18 @@ class TemporaryFolder {
-  mkdir((dir_name + "/data").c_str(), S_IRUSR | S_IWUSR);
-  std::ofstream model_data_output(dir_name + "/data/part-0");
+  mkdir(tfrt::StrCat(dir_name, "/data").c_str(), S_IRUSR | S_IWUSR);
+  std::ofstream model_data_output(tfrt::StrCat(dir_name, "/data/part-0"));
```
nits: `ClinkOneHotEncoderModel::loadCppModel(...)` uses `modelData` as the model data file name. Could we use `modelData` instead of `part-0` as the default file name for consistency here? The `reader-side` could still use `getOnlyFileInDirectory(...)` for robustness.

lindong28 commented on java-lib/src/main/java/org/flinkextended/clink/feature/onehotencoder/ClinkOneHotEncoderModel.java:
```diff
@@ -181,13 +186,22 @@ public void close() throws Exception {
+        OutputStream modelDataOutput =
+                new FileOutputStream(Paths.get(tmpDirStr, "data", "modelData").toFile());
+
+        new ByteArrayEncoder().encode(modelDataBytes, modelDataOutput);
```
If the only reader of the output data, which is the C++ `OneHotEncoderModel::load`, does not need to read byte length, would it be simpler not to output this byte length? Then we can also make the logic here consistent with the C++ `saveMetaDataModelData(...)`.

yunfengzhou-hub replied:
I agree. I'll remove the byte array length information from saved model data file.

lindong28 reviewed (commented):
Thanks for the update. Left some comments below.

lindong28 commented on java-lib/src/main/java/org/flinkextended/clink/util/ByteArrayDecoder.java:
```diff
@@ -38,11 +38,18 @@
-                    int actualLen = inputViewStreamWrapper.read(bytes);
-                    Preconditions.checkArgument(expectedLen == actualLen);
-                    return bytes;
+                    int nRead;
```
The `StreamFormat.Reader::read()` Java doc says:
```
Reads the next record. Returns {@code null} when the input has reached its end.
```

So the reader returned by `ByteArrayDecoder` should be able to return multiple model data (of type `byte[]`) if there are so many values in the given `inputStream`, which could be useful when we re-transmit multiple model data for online learning.

However, it seems that the reader returned here can only support returning at most one model data, because the implementation relies on the end-of-stream to detect end of the model data.

yunfengzhou-hub replied:
You are right. I forgot this point. I'll add this back.

> If the only reader of the output data, which is the C++ OneHotEncoderModel::load, does not need to read byte length, would it be simpler not to output this byte length? Then we can also make the logic here consistent with the C++ saveMetaDataModelData(...).

Then as for the previous comment as above, I think we should still also record the length in C++ programs, because C++ and Java programs should read or write to files with the same format.

yunfengzhou-hub replied:
We have discussed offline that protobuf should be able to record the length of protobuf data in the byte array, so we do not need to explicitly record the length. However, as its document pointed out

> If you want to write multiple messages to a single file or stream, it is up to you to keep track of where one message ends and the next begins. The Protocol Buffer wire format is not self-delimiting, so protocol buffer parsers cannot determine where a message ends on their own. The easiest way to solve this problem is to write the size of each message before you write the message itself. When you read the messages back in, you read the size, then read the bytes into a separate buffer, then parse from that buffer. (If you want to avoid copying bytes to a separate buffer, check out the CodedInputStream class (in both C++ and Java) which can be told to limit reads to a certain number of bytes.)
https://developers.google.com/protocol-buffers/docs/techniques

Protobuf does not have such ability, and recording the size on our own is its recommendation. Thus I believe it is reasonable to make this change in PR.

yunfengzhou-hub commented:
Thanks for the comments. I have updated the PR according to the comments.

yunfengzhou-hub commented:
I have updated the java implementation and tests so that operators' `transform()` method would not work unless permanent storage model data path is provided. On the basis of the modified implementation, I cleaned and reorganized the Java code to improve its readability.

lindong28 commented on java-lib/src/main/java/org/flinkextended/clink/feature/onehotencoder/ClinkOneHotEncoderModel.java:
```diff
@@ -43,42 +44,59 @@
     private final Map<Param<?>, Object> paramMap = new HashMap<>();
-    private Table modelDataTable;
+    private String modelDataPath;
+    private transient Table modelDataTable;
```
My understanding is that we are not sure whether this `transient` is necessary here.

To make code consistent (w.r.t. Flink ML), how about we skip the transient here for now? We can either run test or ask other Flink developers regarding whether this is needed for serializability. And if it does, make this change consistently across those algorithms in Flink ML.

yunfengzhou-hub replied:
It makes sense to skip the transient for now as it only makes slight influence to this PR, that `ClinkOneHotEncoderModel` must use static classes for inner functions and operators, instead of making them member class or anonymous class. I'll remove the `transient` key word here.

lindong28 commented on java-lib/src/main/java/org/flinkextended/clink/feature/onehotencoder/ClinkOneHotEncoderModel.java:
```diff
@@ -43,42 +44,59 @@
+        modelDataTable = inputs[0];
+        // Newly set model data might be inconsistent with that stored in given path, thus the path
+        // must be nullified.
+        modelDataPath = null;
```
modelDataPath will only be non-null if `load(...)` is called.

Is there any scenario where we need to load(...) a model and then use setModelData(...). If not, how about we use `checkState(modelDataPath == null)` here?

yunfengzhou-hub replied:
There seems no reasonable usage of setModelData after the load method. I'll add the check and corresponding javadoc.

lindong28 commented on java-lib/src/main/java/org/flinkextended/clink/feature/onehotencoder/ClinkOneHotEncoderModel.java:
```diff
@@ -94,67 +112,23 @@ public ClinkOneHotEncoderModel() {
+        public Row map(Row row) {
             Row resultRow = new Row(inputCols.length);
             for (int i = 0; i < inputCols.length; i++) {
                 String inputCol = inputCols[i];
```
Would it be better to rename `int number` below to either `int value` or `int inputValue`?

yunfengzhou-hub replied:
I think `inputValue` has the best readability among these choices. I'll make the change this way.

lindong28 commented on java-lib/src/main/java/org/flinkextended/clink/feature/onehotencoder/ClinkOneHotEncoderModel.java:
```diff
@@ -43,42 +44,59 @@
 
     public ClinkOneHotEncoderModel() {
-        initializeMapWithDefaultValues(this.paramMap, this);
+        ParamUtils.initializeMapWithDefaultValues(this.paramMap, this);
```
nits: would it be simpler to replace `this.paramMap` with `paramMap`?

lindong28 commented on java-lib/src/main/java/org/flinkextended/clink/feature/onehotencoder/ClinkOneHotEncoderModel.java:
```diff
@@ -43,42 +44,59 @@
         Preconditions.checkArgument(inputCols.length == outputCols.length);
+        Preconditions.checkNotNull(
+                modelDataPath,
+                "Clink operator's transform() method should not be invoked "
```
nits: The following example error message could be easier to understand than using  double negative statements. And it is probably useful to suggest using save() and load() in the error message.

```
The transform() can only be called after a model is loaded with the path to the model data.
```

lindong28 commented on java-lib/src/main/java/org/flinkextended/clink/feature/onehotencoder/ClinkOneHotEncoderModel.java:
```diff
@@ -43,42 +44,59 @@
     }
 
     @Override
     public Table[] transform(Table... inputs) {
```
Could we also add Java doc here explaining the expected behavior of this method? For example, we can say this method can only be called after the model is loaded with the path to model data), and will throw `IllegalStateException` otherwise.

yunfengzhou-hub replied:
Yes, I agree with this point. What's more, It should be generally applicable to all Clink operators that transform can only be called after the load method. For now I think it is proper to add the Javadoc only to OneHotEncoderModel's transform method, but in future we can create a general interface for all such clink models and place this document there.

lindong28 reviewed (commented):
Thanks for the update. The PR is much more readable now. Left some minor comments below.

yunfengzhou-hub commented:
Thanks for the comments above. I have updated the PR according to these comments.

yunfengzhou-hub committed:
remove duplicate maven dependency

lindong28 reviewed (commented):
Thanks for the update! LGTM.
